lib/rex/proto/ldap/server.rb

```diff
@@ -51,29 +51,32 @@ def write(data)
         #
         # Create LDAP Server
         #
-        # @param lhost [String] Listener address
-        # @param lport [Fixnum] Listener port
-        # @param udp [TrueClass, FalseClass] Listen on UDP socket
-        # @param tcp [TrueClass, FalseClass] Listen on TCP socket
-        # @param ldif [String] LDIF data
+        # @param opts [Hash] Options required to start service
         # @param ctx [Hash] Framework context for sockets
         # @param dblock [Proc] Handler for :dispatch_request flow control interception
         # @param sblock [Proc] Handler for :send_response flow control interception
         #
         # @return [Rex::Proto::LDAP::Server] LDAP Server object
-        def initialize(lhost = '0.0.0.0', lport = 389, udp = true, tcp = true, ldif = nil, comm = nil, ctx = {}, dblock = nil, sblock = nil)
-          @serve_udp = udp
-          @serve_tcp = tcp
+        def initialize(opts = {}, ctx = {}, dblock = nil, sblock = nil)
+          @serve_udp = opts[:udp]
+          @serve_tcp = opts[:tcp]
           @sock_options = {
-            'LocalHost' => lhost,
-            'LocalPort' => lport,
+            'LocalHost' => opts[:bindhost],
+            'LocalPort' => opts[:bindport],
             'Context' => ctx,
-            'Comm' => comm
+            'Comm' => opts[:comm]
           }
-          @ldif = ldif
+          @ldif = opts[:ldif]
           self.listener_thread = nil
           self.dispatch_request_proc = dblock
           self.send_response_proc = sblock
+          @auth_options  = { #check for nill cases
+            challenge: opts[:challenge],
+            domain: opts[:domain],
+            server: opts[:server],
+            dnsname: opts[:dnsname],
+            dnsdomain: opts[:dnsdomain]
+          }
```

Comment:
Again this signature should not have needed to change, the only value possible needed here is that the constructor could accept a `challenge` value to pass into the Auth handler.

Reply:
would like a bit more detail

Reply:
based on https://github.com/JustAnda7/metasploit-framework/pull/9/files#r1344248560

``` ruby
  if auth.instance_of?(Rex::Proto::LDAP::Auth)
    @auth_provider = auth
  else
    raise ArgumentError.new("auth parameter must be an instance of Rex::Proto::LDAP::Auth") unless 
  end
```

Reply:
This could also create a default `Auth` object if the passed parameter is `nil`.